pkg/build/gather-build-info.go

```diff
@@ -0,0 +1,152 @@
+package build
+
+import (
+	"fmt"
+	"github.com/bugsnag/bugsnag-cli/pkg/android"
+	"github.com/bugsnag/bugsnag-cli/pkg/options"
+	"github.com/bugsnag/bugsnag-cli/pkg/utils"
+)
+
+type SourceControl struct {
+	Provider   string `json:"provider,omitempty"`
+	Repository string `json:"repository,omitempty"`
+	Revision   string `json:"revision,omitempty"`
+}
+
+type CreateBuildInfo struct {
+	ApiKey            string            `json:"apiKey,omitempty"`
+	AppVersionCode    string            `json:"appVersionCode,omitempty"`
+	AppBundleVersion  string            `json:"appBundleVersion,omitempty"`
+	SourceControl     SourceControl     `json:"sourceControl,omitempty"`
+	BuilderName       string            `json:"builderName,omitempty"`
+	ReleaseStage      string            `json:"releaseStage,omitempty"`
+	AppVersion        string            `json:"appVersion,omitempty"`
+	AutoAssignRelease *bool             `json:"autoAssignRelease,omitempty"`
+	MetaData          map[string]string `json:"metadata,omitempty"`
+}
+
+func (opts CreateBuildInfo) Override(base CreateBuildInfo) CreateBuildInfo {
+	return CreateBuildInfo{
+		ApiKey:           utils.ThisOrThat(opts.ApiKey, base.ApiKey).(string),
+		AppVersionCode:   utils.ThisOrThat(opts.AppVersionCode, base.AppVersionCode).(string),
+		AppBundleVersion: utils.ThisOrThat(opts.AppBundleVersion, base.AppBundleVersion).(string),
+		SourceControl: SourceControl{
+			Provider:   utils.ThisOrThat(opts.SourceControl.Provider, base.SourceControl.Provider).(string),
+			Repository: utils.ThisOrThat(opts.SourceControl.Repository, base.SourceControl.Repository).(string),
+			Revision:   utils.ThisOrThat(opts.SourceControl.Revision, base.SourceControl.Revision).(string),
+		},
+		BuilderName:       utils.ThisOrThat(opts.BuilderName, base.BuilderName).(string),
+		ReleaseStage:      utils.ThisOrThat(opts.ReleaseStage, base.ReleaseStage).(string),
+		AppVersion:        utils.ThisOrThat(opts.AppVersion, base.AppVersion).(string),
+		AutoAssignRelease: utils.ThisOrThatBool(opts.AutoAssignRelease, base.AutoAssignRelease),
+		MetaData:          utils.ThisOrThat(opts.MetaData, base.MetaData).(map[string]string),
+	}
+}
+
+func (opts CreateBuildInfo) Validate() error {
+	if opts.ApiKey == "" {
+		return fmt.Errorf("no API key provided")
+	}
+
+	if opts.AppVersion == "" {
+		return fmt.Errorf("Missing app version, please provide this via the command line options")
+	}
+
+	if opts.SourceControl.Repository == "" || opts.SourceControl.Revision == "" {
+		return fmt.Errorf("Missing Source Control Repository or Revision")
+	}
```

Comment:
Are these fields definitely required for the Build API or can the whole of `sourceControl` be omitted?

If this is true, we should split the validation and resulting error messages.

Reply:
According to [this](https://bugsnagbuildapi.docs.apiary.io/#reference/0/build/notify-of-a-build). `repository` and `revision` are. Provider looks to be inferred on the API side which is why there is no validation within the CLI on that option.

Reply:
Would you mind trying a few combinations against the API? I'm a little doubtful if that is strictly the case as I'd have thought you'd be able to create a release without source control information. Perhaps they're required if you want to provide any `sourceControl` fields?

Reply:
My mistake. You are correct. They're only required if you pass the `sourceControl` element

Reply:
Discussed IRL - we'll let the API do the validation.